rfcs/XXXX-gating/README.md

```diff
@@ -0,0 +1,301 @@
+to production by reviewing and approving the proposed changes in a collaborative manner with pull request.
+Once a pull request is merged onto a branch that defines the desired state of the production system,
+Flux kicks off the reconciliation process.
+
+There are situations when users want to have a gating mechanism after the desired state changes are merged in Git:
+
+- Manual approval of container image updates (e.g. https://github.com/fluxcd/flux2/discussions/870)
+- Manual approval of infrastructure upgrades (e.g. https://github.com/fluxcd/flux2/issues/959)
+- Maintenance window (e.g. https://github.com/fluxcd/flux2/discussions/1004)
+- Planned releases
+- No Deploy Friday
+
+### Goals
+
+- Offer a dedicated API for defining time-based gates in a declarative manner.
+- Introduce a `gating-controller` in the Flux suite that manages the `Gate` objects.
+- Extend the current Flux APIs and controllers to support gating.
+
+### Non-Goals
+
+<!--
+What is out of scope for this RFC? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+In order to support manual gating, Flux could be extended with a dedicated API and controller
+that would allow users to define `Gate` objects and perform operations like `open` and `close`.
+
+A `Gate` object could be referenced in sources (Buckets, Git, Helm, OCI Repositories)
+and syncs (Kustomizations, HelmReleases, ImageUpdateAutomation)
+to block the reconciliation until the gate is opened.
+
+A `Gate` can be opened or closed by annotating the object with a timestamp or by
+calling a specific webhook receiver exposed by notification-controller.
+
+A `Gate` can be configured to automatically close or open based on a time window defined in the `Gate` spec.
+
+The `Gate` API would replace Flagger's current
+[manual gating mechanism](https://docs.flagger.app/usage/webhooks#manual-gating).
+
+### User Stories
+
+#### Story 1
+
+> As a member of the SRE team, I want to allow deployments to happen only
+> in a particular time frame of my own choosing.
+
+Define a gate that automatically closes after 1h from the time it has been opened:
+
+```yaml
+apiVersion: gating.toolkit.fluxcd.io/v1alpha1
+kind: Gate
+metadata:
+  name: sre-approval
+  namespace: flux-system
+spec:
+  interval: 30s
+  default: closed
+  window: 1h
+```
+
+When the gate is created in-cluster, the `gating-controller` uses `spec.default` to set the `Opened` condition:
+
+```yaml
+apiVersion: gating.toolkit.fluxcd.io/v1alpha1
+kind: Gate
+metadata:
+  name: sre-approval
+  namespace: flux-system
+status:
+  conditions:
+    - lastTransitionTime: "2021-03-26T10:09:26Z"
+      message: "Gate closed by default"
+      reason: ReconciliationSucceeded
+      status: "False"
+      type: Opened
+```
+
+While the gate is closed, all the objects that reference it will wait for an approval:
+
+```yaml
+apiVersion: kustomize.toolkit.fluxcd.io/v1beta1
+kind: Kustomization
```

Comment:
I propose gates are to be applied at sources only instead of `Kustomization` and `HelmRelease` objects, as otherwise may face the non-trivial challenges below:
- Incompatible desired state when `Kustomization` / `HelmRelease` objects are gated whilst others are not, specially when they depend on the same Flux source that is not gated.
  - Example 1: `Kustomization` objects `app2` and `app1` depends on `GitRepository` `team-source`. A change is merged into the repository and reconciled for object `team-source`, however `app2` has a close Gate and `app1` does not. In this case the desired state at both Git repository and the Flux source in the cluster are valid, but only one Kustomizations is being applied leading to an invalid observed state in the cluster. Same applies if both Kustomizations are behind a Gate which has its state changed before the second object is reconcilied (e.g. example 2 below).
- Enforcing the desired state which was allowed **before the Gate closing** (User Story 2) will be difficult to manage from any point beyond the source, which could lead to the cluster to be in an invalid state.
  - Example 1: user manually deletes Kubernetes object managed in the hope Flux will re-create it.
  - Example 2: two kustomizations (`parent` and `child`) having the same Gate, the latter has a `dependsOn` pointing to the former. If the Gate closes after `parent` but before `child` is applied, the cluster observed state may be invalid (considering the desired state at source).